tests/Messenger/Transport/DsnParserTest.php

```diff
@@ -12,26 +12,196 @@ class DsnParserTest extends TestCase
+     */
+    public static function provideInvalidConfigurations(): iterable
+    {
+        yield 'unparseable DSN' => [
+            'dsn' => 'http://?',
+            'options' => [],
+            'error' =>
+                'Invalid Azure Service Bus DSN for the "my-transport" transport. ' .
+                'It must be in the following format: azure://SharedAccessKeyName:SharedAccessKey@namespace',
+            'code' => 1643988474,
+        ];
+
+        yield 'wrong scheme' => [
+            'dsn' => 'http://SharedAccessKeyName:SharedAccessKey@namespace',
+            'options' => [],
+            'error' =>
+                'Invalid Azure Service Bus DSN for the "my-transport" transport. ' .
+                'It must be in the following format: azure://SharedAccessKeyName:SharedAccessKey@namespace',
+            'code' => 1643988474,
+        ];
+
+        yield 'unknown option' => [
+            'dsn' => 'azure://SharedAccessKeyName:SharedAccessKey@namespace',
+            'options' => ['foo' => 'bar'],
+            'error' =>
+                'Unknown option found: [foo]. ' .
+                'Allowed options are [shared_access_key_name, shared_access_key, namespace, entity_path, ' .
+                'subscription, token_expiry, receive_mode].',
+        ];
+
+        yield 'unknown query param' => [
+            'dsn' => 'azure://SharedAccessKeyName:SharedAccessKey@namespace?foo=bar',
+            'options' => [],
+            'error' =>
+                'Unknown option found in DSN: [foo]. ' .
+                'Allowed options are [shared_access_key_name, shared_access_key, namespace, entity_path, ' .
+                'subscription, token_expiry, receive_mode].',
+        ];
+
+        yield 'missing entity_path' => [
+            'dsn' => 'azure://SharedAccessKeyName:SharedAccessKey@namespace',
+            'options' => [],
+            'error' => 'Missing entity_path (queue or topic) for the "my-transport" transport.',
+            'code' => 1643989596,
+        ];
+
+        yield 'invalid receive_mode (in options)' => [
+            'dsn' => 'azure://SharedAccessKeyName:SharedAccessKey@namespace',
+            'options' => [
+                'entity_path' => 'foo',
+                'receive_mode' => 'foo',
+            ],
+            'error' => 'Invalid "foo" receive_mode for the "my-transport" transport. It must be one of: ' .
+                'peek-lock, receive-and-delete.',
+            'code' => 1643994036,
+        ];
+
+        yield 'invalid receive_mode (in DNS)' => [
+            'dsn' => 'azure://SharedAccessKeyName:SharedAccessKey@namespace?receive_mode=foo',
+            'options' => [
+                'entity_path' => 'foo',
+            ],
+            'error' => 'Invalid "foo" receive_mode for the "my-transport" transport. It must be one of: ' .
+                'peek-lock, receive-and-delete.',
+            'code' => 1643994036,
+        ];
+    }
```

Comment:
As you removed the RegEx for DSN parsing, we can actually configure a transport without key/key name. This would be fixed with the change I suggested in `pickOption()`, we can add the test cases:
```php
yield 'missing key name' => [
    'dsn' => 'azure://:SharedAccessKey@namespace?receive_mode=peek-lock',
    'options' => [
        'entity_path' => 'foo',
    ],
    'error' => 'Missing shared_access_key_name for the "my-transport" transport.',
    'code' => 1702724695,
];

yield 'missing key' => [
    'dsn' => 'azure://SharedAccessKeyName@namespace?receive_mode=peek-lock',
    'options' => [
        'entity_path' => 'foo',
    ],
    'error' => 'Missing shared_access_key for the "my-transport" transport.',
    'code' => 1702724695,
];
```
I did not add a test case for the `namespace` option as I don't think it is doable.